tools/run_tests/run_xds_tests.py

```diff
@@ -1782,6 +1815,90 @@ def _delay(pct):
         set_validate_for_proxyless(gcp, True)
 
 
+def test_csds(gcp, original_backend_service, instance_group, server_uri):
+    logger.info('Running csds')
+
+    logger.info('waiting for original backends to become healthy')
+    wait_for_healthy_backends(gcp, original_backend_service, instance_group)
+
+    logger.info('start sending traffic')
+    configure_client(
+        rpc_types=[messages_pb2.ClientConfigureRequest.RpcType.UNARY_CALL])
+
+    # Test case timeout: 5 minutes
+    deadline = time.time() + 300
+    cnt = 0
+    while time.time() <= deadline:
+        client_config = get_client_xds_config_dump()
+        logger.info('received xDS config %s', json.dumps(client_config,
+                                                         indent=2))
+        if client_config is not None:
+            # Got the xDS config dump, now validate it
+            ok = True
+            try:
+                if client_config['node']['locality']['zone'] != args.zone:
```

Comment:
Are there consts you can use (e.g. from the generated proto code) instead of these strings?  Or: why convert to JSON instead of returning the data as a proto message?

Reply:
CSDS output is basically a proto message with lots of layers of `google.protobuf.Any`. If we use ProtoBuf Python to unpack each layer, then test it's value, then dig into its children, we will need much more LOC.

If the parsing of CSDS output failed, it would be a pretty obvious error to debug.

So, for the sake of the simplifying the test logic, I think it would be better to validate the CSDS output as JSON instead of ProtoBuf message.

Reply:
SGTM